# Lesson 01 ... Welcome to pandas

We will start with importing some libraries we need and then play with some data to understand basic python commands. What data shall we work with? Well, let us pull down some data on criminal incidences that were reported.

First we install a particular library called `pandas` and in the command that follows, note that `pd` is just the alias that pandas assumes so that we can type `pd` and have all the `pandas` commands at our disposal.

In [1]:
import pandas as pd 
import numpy as np

The crime incident reports data are [available here]("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv") and span multiple years so we may end up working only with 2019 data but for now we proceed by gathering everything.

In the command below, the key part is `pd.read_csv()` and inside it is the URL for the comma-separated variable file. Once the file is downloaded by `pandas` we are saving it in python with the name `df` 

Note that data-sets, data-files are usually referred to as a `data-frame` in python and hence the alias of `df`.


In [2]:
# df = pd.read_csv("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/12cb3883-56f5-47de-afa5-3b1cf61b257b/download/tmpayw7hysb.csv")

df = pd.read_csv("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/54e0eb09-c012-429f-90e8-8c5f2694ce27/download/crime-incident-reports-2019.csv")

/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3147: DtypeWarning: Columns (0,2,6,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Let us look at the first 5 rows of data to get a feel for the layout. The command is `.head(5)`

In [3]:
df.head(5)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP   OFFENSE_DESCRIPTION  \
0       TESTTEST2           423                NaN  ASSAULT - AGGRAVATED   
1       TESTTEST2           423                NaN  ASSAULT - AGGRAVATED   
2       TESTTEST2           423                NaN  ASSAULT - AGGRAVATED   
3       TESTTEST2           423                NaN  ASSAULT - AGGRAVATED   
4       TESTTEST2           423                NaN  ASSAULT - AGGRAVATED   

   DISTRICT REPORTING_AREA SHOOTING     OCCURRED_ON_DATE  YEAR  MONTH  \
0  External                       0  2019-10-16 00:00:00  2019     10   
1  External                       0  2019-10-16 00:00:00  2019     10   
2  External                       0  2019-10-16 00:00:00  2019     10   
3  External                       0  2019-10-16 00:00:00  2019     10   
4  External                       0  2019-10-16 00:00:00  2019     10   

  DAY_OF_WEEK  HOUR UCR_PART        STREET  Lat  Long  \
0   Wednesday     0      NaN  RIVERVIEW DR  NaN   NaN   
1   Wednesday     0      NaN  RIVERVIEW DR  NaN   NaN   
2   Wednesday     0      NaN  RIVERVIEW DR  NaN   NaN   
3   Wednesday     0      NaN  RIVERVIEW DR  NaN   NaN   
4   Wednesday     0      NaN  RIVERVIEW DR  NaN   NaN   

                   Location  
0  (0.00000000, 0.00000000)  
1  (0.00000000, 0.00000000)  
2  (0.00000000, 0.00000000)  
3  (0.00000000, 0.00000000)  
4  (0.00000000, 0.00000000)

What about the last 10 rows of the data?

In [4]:
df.tail(10)

INCIDENT_NUMBER  OFFENSE_CODE OFFENSE_CODE_GROUP  \
784761       102091671          2647                NaN   
784762       102091671          2647                NaN   
784763       102091671          2647                NaN   
784764       102091671          2647                NaN   
784765       102091671          2647                NaN   
784766       102091671          2647                NaN   
784767       102091671          2647                NaN   
784768       102091671          2647                NaN   
784769       102091671          2647                NaN   
784770       102091671          2647                NaN   

              OFFENSE_DESCRIPTION DISTRICT REPORTING_AREA SHOOTING  \
784761  THREATS TO DO BODILY HARM       B3            417        0   
784762  THREATS TO DO BODILY HARM       B3            417        0   
784763  THREATS TO DO BODILY HARM       B3            417        0   
784764  THREATS TO DO BODILY HARM       B3            417        0   
784765  THREATS TO DO BODILY HARM       B3            417        0   
784766  THREATS TO DO BODILY HARM       B3            417        0   
784767  THREATS TO DO BODILY HARM       B3            417        0   
784768  THREATS TO DO BODILY HARM       B3            417        0   
784769  THREATS TO DO BODILY HARM       B3            417        0   
784770  THREATS TO DO BODILY HARM       B3            417        0   

           OCCURRED_ON_DATE  YEAR  MONTH DAY_OF_WEEK  HOUR UCR_PART   STREET  \
784761  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784762  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784763  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784764  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784765  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784766  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784767  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784768  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784769  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   
784770  2019-11-12 12:00:00  2019     11     Tuesday    12      NaN  MORA ST   

              Lat       Long                     Location  
784761  42.282081 -71.073648  (42.28208100, -71.07364800)  
784762  42.282081 -71.073648  (42.28208100, -71.07364800)  
784763  42.282081 -71.073648  (42.28208100, -71.07364800)  
784764  42.282081 -71.073648  (42.28208100, -71.07364800)  
784765  42.282081 -71.073648  (42.28208100, -71.07364800)  
784766  42.282081 -71.073648  (42.28208100, -71.07364800)  
784767  42.282081 -71.073648  (42.28208100, -71.07364800)  
784768  42.282081 -71.073648  (42.28208100, -71.07364800)  
784769  42.282081 -71.073648  (42.28208100, -71.07364800)  
784770  42.282081 -71.073648  (42.28208100, -71.07364800)

Let us look at the contents of the data-frame ... 

| Column Name | Description |
| :--          | :--- |
| [incident_num] [varchar](20) NOT NULL, | Internal BPD report number |
| [offense_code][varchar](25) NULL,| Numerical code of offense description |
| [Offense_Code_Group_Description][varchar](80) NULL, | Internal categorization of [offense_description] |
| [Offense_Description][varchar](80) NULL, | Primary descriptor of incident |
| [district] [varchar](10) NULL,| What district the crime was reported in |
| [reporting_area] [varchar](10) NULL, | RA number associated with the where the crime was reported from. |
| [shooting][char] (1) NULL, | Indicated a shooting took place. |
| [occurred_on] [datetime2](7) NULL, | Earliest date and time the incident could have taken place |
| [UCR_Part] [varchar](25) NULL,| Universal Crime Reporting Part number (1,2, 3) |
| [street] [varchar](50) NULL,| Street name the incident took place |


Offense Codes are [available here](https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx)

We could also look at the offense codes by reading them in as a data-frame. This is an Excel file so we will have to switch to `.read_excel()`


In [6]:
offense_codes = pd.read_excel("https://data.boston.gov/dataset/6220d948-eae2-4e4b-8723-2dc8e67722a3/resource/3aeccf51-a231-4555-ba21-74572b4c33d6/download/rmsoffensecodes.xlsx")

In [7]:
print(offense_codes)

     CODE                                       NAME
0     612           LARCENY PURSE SNATCH - NO FORCE 
1     613                        LARCENY SHOPLIFTING
2     615    LARCENY THEFT OF MV PARTS & ACCESSORIES
3    1731                                     INCEST
4    3111                  LICENSE PREMISE VIOLATION
..    ...                                        ...
571  1806  DRUGS - CLASS B TRAFFICKING OVER 18 GRAMS
572  1807  DRUGS - CLASS D TRAFFICKING OVER 50 GRAMS
573  1610    HUMAN TRAFFICKING - COMMERCIAL SEX ACTS
574  2010                              HOME INVASION
575  1620  HUMAN TRAFFICKING - INVOLUNTARY SERVITUDE

[576 rows x 2 columns]


The next step would be to see how many data points we have, and what the minimum, maximum values, what is the average, etc. This can be done with `.describe()`

In [8]:
df.describe()

OFFENSE_CODE      YEAR          MONTH           HOUR            Lat  \
count  784771.000000  784771.0  784771.000000  784771.000000  767540.000000   
mean     2377.351602    2019.0      10.390184      12.860722      42.314448   
std      1193.299500       0.0       2.030411       6.567807       0.532789   
min       111.000000    2019.0       1.000000       0.000000      -1.000000   
25%       802.000000    2019.0      10.000000       9.000000      42.295353   
50%      3006.000000    2019.0      11.000000      14.000000      42.324694   
75%      3301.000000    2019.0      12.000000      18.000000      42.348406   
max      3831.000000    2019.0      12.000000      23.000000      42.395042   

                Long  
count  767540.000000  
mean      -71.073258  
std         0.868824  
min       -71.177457  
25%       -71.098172  
50%       -71.078610  
75%       -71.061776  
max         0.000000

By default the command will report the values with decimals but we may not want that. Decimals can be `rounded` or removed altogether as shown below.

In [9]:
df.describe().round(2)

OFFENSE_CODE      YEAR      MONTH       HOUR        Lat       Long
count     784771.00  784771.0  784771.00  784771.00  767540.00  767540.00
mean        2377.35    2019.0      10.39      12.86      42.31     -71.07
std         1193.30       0.0       2.03       6.57       0.53       0.87
min          111.00    2019.0       1.00       0.00      -1.00     -71.18
25%          802.00    2019.0      10.00       9.00      42.30     -71.10
50%         3006.00    2019.0      11.00      14.00      42.32     -71.08
75%         3301.00    2019.0      12.00      18.00      42.35     -71.06
max         3831.00    2019.0      12.00      23.00      42.40       0.00

In [10]:
df.describe().round(0)

OFFENSE_CODE      YEAR     MONTH      HOUR       Lat      Long
count      784771.0  784771.0  784771.0  784771.0  767540.0  767540.0
mean         2377.0    2019.0      10.0      13.0      42.0     -71.0
std          1193.0       0.0       2.0       7.0       1.0       1.0
min           111.0    2019.0       1.0       0.0      -1.0     -71.0
25%           802.0    2019.0      10.0       9.0      42.0     -71.0
50%          3006.0    2019.0      11.0      14.0      42.0     -71.0
75%          3301.0    2019.0      12.0      18.0      42.0     -71.0
max          3831.0    2019.0      12.0      23.0      42.0       0.0

Note a few things here. 

* We have a total of 515082 incidents logged. But the latitude and longitude are availale for no more than 485909 incidents. 


Say we want to restrict the dataframe just to 2020. How can we do that?

df20 = df[ df['YEAR'] == 2020 ]

Notice the sequence here `dataframe[ dataframe['column-name'] == somevalue ]` & pay attention to the double equal sign `==` which is a strict equality. 

At this point we might be curious to know what types of offenses are most often reported? Before we that, however, let us also see how many unique values of OFFENSE_CODE are there

In [15]:
df['OFFENSE_CODE'].nunique()

194

In [16]:
df['OFFENSE_DESCRIPTION'].nunique()

222

In [19]:
off_des = df['OFFENSE_DESCRIPTION'] 

off_des.to_csv('~/Downloads/off_desc.csv', index=False)

In [14]:
df['OFFENSE_CODE'].value_counts()

3301    63919
3115    55067
801     47641
3831    40144
3005    31233
        ...  
121         1
2672        1
1502        1
2663        1
2636        1
Name: OFFENSE_CODE, Length: 194, dtype: int64

So code 3301 leads with 6234 reports in 2020, followed by code 3115, then 801, then 3005, and then 3831. Code 3005 is missing from their list so we have no idea what it is!! That is a crime in itself.   

In [ ]:
# Just another way to accomplish the same thing but in a more complicated way.

df.groupby('OFFENSE_CODE')['OFFENSE_CODE'].count().reset_index(name='count').sort_values(['count'], ascending = False) 

Let us focus in on these verbal disputes. We will do so by creating a new dataframe that is only for OFFENSE_CODE 3301.

In [ ]:
dfverbal = df20[ df20['OFFENSE_CODE'] == 3301 ]

In [ ]:
# Now we see this dataframe just to check

dfverbal

Which days of the week have more verbal disputes?

In [ ]:
dfverbal['DAY_OF_WEEK'].value_counts()

Which hour, which streets have the most verbal disputes?

In [ ]:
dfverbal['HOUR'].value_counts()

In [ ]:
dfverbal['STREET'].value_counts()

What districts are the worst?

In [ ]:
dfverbal['DISTRICT'].value_counts()

Lookup districts C11, B3, and B2 ... what areas are these?

# Practice Task 01

Pick another data-set from `data.boston.gov` and go through the same commands, picking some interesting element of the dataframe to explore

 ## What if we have local data, sitting in a folder called, data?¶

In [ ]:
bm15 = pd.read_csv("data/marathon_results_2015.csv")
bm15.head()

In [ ]:
bm16 = pd.read_csv("data/marathon_results_2016.csv")
bm16.head()

In [ ]:
bm17 = pd.read_csv("data/marathon_results_2017.csv")
bm17.head()

Notice the extra first column in `bm15` and `bm17` ... these should be dropped. 

We can then create an indicator for which year the race data are from, and then combine all three data-frames so that we have a single file. When we do this we will write it out as a `csv` file too.

In [ ]:
bm15 = bm15.drop(columns = ['Unnamed: 0', 'Unnamed: 9'])
bm15.head()

In [ ]:
bm17 = bm17.drop(columns = ['Unnamed: 0', 'Unnamed: 9'])
bm17.head()

In [ ]:
bm15['Year'] = '2015'
bm16['Year'] = '2016'
bm17['Year'] = '2017'

Now we bind the rows (i.e., stack one data-frame on top of the other with `concat( [dfi, dfj, dfk] ) 

In [ ]:
bm_df = pd.concat([bm15, bm16, bm17])

In [ ]:
bm_df.shape 

# How many rows and how many columns do we have? 
# This is a good step to check if we stacked the data-frames correctly or not

### How many runners per year?

In [ ]:
bm_df['Year'].value_counts()

### How many Male/Female runners per year?

In [ ]:
bm_df.groupby('M/F')['Year'].value_counts()

In [ ]:
# The proportion of Male/Female runners, by Year

bm_df.groupby('Year')['M/F'].value_counts(normalize = True)